src/primitives.cc

```diff
@@ -121,20 +121,20 @@ class PrimitiveNode : public LeafNode
  */
 Value PrimitiveModule::lookup_radius(const Context &ctx, const Location &loc, const std::string &diameter_var, const std::string &radius_var) const
 {
-	auto d = ctx.lookup_variable(diameter_var, true);
-	auto r = ctx.lookup_variable(radius_var, true);
-	const auto r_defined = (r->type() == Value::ValueType::NUMBER);
+	const auto &d = ctx.lookup_variable(diameter_var, true);
+	const auto &r = ctx.lookup_variable(radius_var, true);
+	const auto r_defined = (r.type() == Value::ValueType::NUMBER);
 	
-	if (d->type() == Value::ValueType::NUMBER) {
+	if (d.type() == Value::ValueType::NUMBER) {
 		if (r_defined) {
 			std::string locStr = loc.toRelativeString(ctx.documentPath());
 			PRINTB("WARNING: Ignoring radius variable '%s' as diameter '%s' is defined too, %s", radius_var % diameter_var % locStr);
 		}
-		return {d->toDouble() / 2.0};
+		return Value(d.toDouble() / 2.0);
 	} else if (r_defined) {
-		return *r;
+		return r.clone();
 	} else {
-		return Value::undefined;
+		return Value();
```

Comment:
OK, so this is another issue with the move vs copy change.  Places that return a const reference still use `Value::undefined`  Otherwise if its return by value, since `Value::undefined` is const, it requires a copy or a new construction.  So in these places I could explicitly call `Value::undefined.clone()` but that seemed weird and messy looking.  I actually toyed with the idea of using a static function `Value::undefined()` to return a newly constructed undefined value, but then the const ref instances would need to change as I don't think we can have redefined member function and static member data with the same name.   What do you think about implementing both with slightly different naming, maybe the call the static member function `Value::undefined()`, and call the const ref `Value::undef`, to match with scad nomenclature?

Reply:
I've been continuing work on this PR lately, in t-paul's value-rework-2 branch.  I have addressed this by adding a static `Value::undef()`  member function for cases where undefined is needed by-value (I guess I ended up doing the opposite naming of what I proposed above).

And for extra completeness, I made `Value()` default constructor as private, to enforce being explicit about undefined values. Values should be immutable anyways so I figured it didn't make much sense to allow default constructing one then change it later.
This had a side effect of pointing out another place where Values were being inadvertently default constructed:  
when `unordered_map<value>::operator[]` is used to assign a value; If it requires an insert, then a value is default constructed, and this was happening for `Context`'s `ValueMap` typedef in a couple places.  
In C++17 there is `unordered_map::insert_or_assign` function which seems to be a way to workaround this default construction, so I've made a function which I think is more-or-less equivalent in operation, and put ValueMap into its own wrapper class with its own header `valuemap.h` instead of a typedef under `context.h`.  I also was able to simplify a few other operations and duplicated `unordered_map::find` calls while looking over the calling points for this class.
Lastly, `dxfdim.cc` used a similar value map cache which now is able to re-use class `ValueMap` too.